src/lua/uri.c

```diff
@@ -0,0 +1,350 @@
+		assert(rc == 0 || !diag_is_empty(diag_get()));
+		lua_pop(L, 1);
+	}
+	return rc;
+}
+
+/**
+ * Returns the type of the field, which located at the given valid
+ * @a index in the table which located at the given valid @a table_idx.
+ */
+static int
+field_type(struct lua_State *L, int table_idx, int index)
+{
+	assert(lua_type(L, table_idx) == LUA_TTABLE);
+	lua_rawgeti(L, table_idx, index);
+	int rc = lua_type(L, -1);
+	lua_pop(L, 1);
+	return rc;
+}
+
+/**
+ * Check if there is a field with the name @a name in the table,
+ * which located which located at the given valid @a idx, which
+ * should be positive value.
+ */
+static bool
+is_field_present(struct lua_State *L, int idx, const char *name)
+{
+	assert(lua_type(L, idx) == LUA_TTABLE);
+	lua_pushstring(L, name);
+	lua_rawget(L, idx);
+	bool field_is_present = (lua_type(L, -1) != LUA_TNIL);
+	lua_pop(L, 1);
+	return field_is_present;
+}
+
+/**
+ * Create @a uri from the table, which located at the given valid @a idx,
+ * which should be positive value.
+ */
+static int
+uri_create_from_lua_table(struct lua_State *L, int idx, struct uri *uri)
+{
+	assert(lua_type(L, idx) == LUA_TTABLE);
+	/* There should be exactly one URI in the table */
+	int size = lua_objlen(L, idx);
+	int uri_count = size + is_field_present(L, idx, "uri");
+	if (uri_count != 1) {
+		diag_set(IllegalParams, "Invalid URI table: "
+			 "expected {uri = string, params = table} "
+			 "or {string, params = table}");
+		return -1;
+	}
+	/* Table "default_params" is not allowed for single URI */
+	if (is_field_present(L, idx, "default_params")) {
+		diag_set(IllegalParams, "Default URI query parameters are "
+			 "not allowed for single URI");
+		return -1;
+	}
+	int rc = 0;
+	if (size == 1) {
+		lua_rawgeti(L, idx, 1);
+	} else {
+		lua_pushstring(L, "uri");
+		lua_rawget(L, idx);
+	}
+	const char *uristr = lua_tostring(L, -1);
+	if (uristr != NULL) {
+		rc = uri_create(uri, uristr);
+		if (rc != 0) {
+			diag_set(IllegalParams, "Incorrect URI: expected "
+				 "host:service or /unix.socket");
+		}
+	} else {
+		diag_set(IllegalParams, "Incorrect type for URI in nested "
+			 "table: should be string, number");
+		rc = -1;
+	}
+	lua_pop(L, 1);
+	if (rc != 0)
+		return rc;
+	lua_pushstring(L, "params");
+	lua_rawget(L, idx);
+	rc = uri_add_params_from_lua(uri, L, true);
+	lua_pop(L, 1);
+	if (rc != 0)
+		uri_destroy(uri);
+	return rc;
+
+}
+
+/**
+ * Create @a uri from the value at the given valid @a idx, which
+ * should be positive value.
+ */
+static int
+luaT_uri_create(struct lua_State *L, int idx, struct uri *uri)
+{
+	int rc = 0;
+	if (lua_isstring(L, idx)) {
+		rc = uri_create(uri, lua_tostring(L, idx));
+		if (rc != 0) {
+			diag_set(IllegalParams, "Incorrect URI: "
+				 "expected host:service or "
+				 "/unix.socket");
+		}
+	} else if (lua_istable(L, idx)) {
+		rc = uri_create_from_lua_table(L, idx, uri);
+	} else if (lua_isnil(L, idx)) {
+		uri_create(uri, NULL);
```

Comment:
What's the point handling nil here? Shouldn't it raise an error in this case, like it does for other types below? Same question for `luaT_uri_create_set` - why do we need to handle nil?

Reply:
As for `box.cfg.listen` as for `box.cfg.replication` we can pass `nil` or `""` value, which all leds to a nil value for this option.
We should return empty URI set for this case because in iproto it's means stop listen not error.

Reply:
Okay, I see.